scripts/update-spec-links.sh

```diff
@@ -0,0 +1,718 @@
+link_rx='^\+\+\+ https://github\.com/provenance-io/provenance.*/proto/.*\.proto'
+link_rx_esc="$( sed -E 's|\\|\\\\|g;' <<< "$link_rx" )"
+declare files=()
+
+if [[ "${#files_in[@]}" -eq '0' && "${#dirs_in[@]}" -eq '0' ]]; then
+  dirs_in+=( '.' )
+fi
+
+i=0
+for file in "${files_in[@]}"; do
+  i=$(( i + 1 ))
+  [[ -n "$verbose" ]] && printf '[%d/%d]: Checking file: %s ... ' "$i" "${#files_in[@]}" "$file"
+  if grep -E --no-messages --binary-file=without-match --silent "$link_rx" "$file" > /dev/null 2>&1; then
+    files+=( "$file" )
+    [[ -n "$verbose" ]] && printf 'File has links.\n'
+  else
+    if [[ -n "$verbose" ]]; then
+      printf 'No links found.\n'
+    elif [[ -z "$quiet" ]]; then
+     printf 'File does not have any links: %s\n' "$file"
+   fi
+  fi
+done
+
+i=0
+for folder in "${dirs_in[@]}"; do
+  i=$(( i + 1 ))
+  [[ -n "$verbose" ]] && printf '[%d/%d]: Checking directory for files with links: %s ... ' "$i" "${#dirs_in[@]}" "$folder"
+  declare new_files=()
+  set -o noglob
+  IFS=$'\n' new_files+=( $( grep -E --files-with-matches --recursive --no-messages --binary-file=without-match "$link_rx" "$folder" | sort ) )
+  set +o noglob
+  if [[ "${#new_files[@]}" -ne '0' ]]; then
+    [[ -n "$verbose" ]] && printf 'Found %d file(s) with links:\n' "${#new_files[@]}"
+    j=0
+    for file in "${new_files[@]}"; do
+      j=$(( j + 1 ))
+      [[ -n "$verbose" ]] && printf '[%d/%d|%d/%d]:   %s' "$i" "${#dirs_in[@]}" "$j" "${#new_files[@]}" "$file"
+      if [[ "$file" =~ \.md$ ]]; then
+        files+=( "$file" )
+        [[ -n "$verbose" ]] && printf '\n'
+      else
+        [[ -n "$verbose" ]] && printf ' (ignored, not a markdown file)\n'
+      fi
+    done
+  else
+    if [[ -n "$verbose" ]]; then
+      printf 'None found.\n'
+    elif [[ -z "$quiet" ]]; then
+      printf 'Directory does not have any files with links: %s\n' "$folder"
+    fi
+  fi
+done
+
+if [[ "${#files[@]}" -eq '0' ]]; then
+  printf 'No files found to update.\n'
+  exit 0
+fi
+
+[[ -z "$quiet" ]] && printf 'Updating links in %d file(s).\n' "${#files[@]}"
+[[ -n "$verbose" ]] && printf '  %s\n' "${files[@]}"
+
+########################################################################################################################
+###########################################  Identify proto files involved.  ###########################################
+########################################################################################################################
+
+[[ -z "$quiet" ]] && printf 'Identifying proto files linked to from %d markdown files.\n' "${#files[@]}"
+
+declare protos_linked=()
+i=0
+for file in "${files[@]}"; do
+  i=$(( i + 1 ))
+  [[ -n "$verbose" ]] && printf '[%d/%d]: Checking %s ... ' "$i" "${#files[@]}" "$file"
+  declare new_proto_files=()
+  set -o noglob
+  IFS=$'\n' new_proto_files+=( $( grep -E "$link_rx" "$file" | sed 's|^.*/proto/|proto/|; s/#.*$//;' | sort -u ) )
```

Comment:
`mapfile` is not something we can reliably expect environments to have.
Similarly, not all systems have a `read` utility that recognizes the `-a` flag.

Reply:
A `while read` loop is significantly slower than just about any alternative. And it's noticeably slower too. Additionally, it will just skip the last line if there's no ending newline. Various environments will do different things with respects to inclusion of a final newline in the output of various commands, though, so you can't just append a newline to that output because then you might be adding a blank line.

The way I have it consistently and quickly does exactly what's needed.

What is wrong with what I have?